File: 백준/Gold/14284. 간선 이어가기 2/간선 이어가기 2.py
from heapq import heappush, heappop
import sys
input = sys.stdin.readline
INF = int(1e9)

def dijkstra(start, graph, n):
    dp = [INF] * (n + 1)
    q = [(0, start)]
    dp[start] = 0
    while q:
        cur_cost, cur_node = heappop(q)
        if cur_cost > dp[cur_node]:
            continue
        for neighbor, cost in graph[cur_node]:
            new_cost = cur_cost + cost
            if new_cost < dp[neighbor]:
                dp[neighbor] = new_cost
                heappush(q, (new_cost, neighbor))
    return dp
    
    
def main():
    v, e = map(int, input().split())
    graph = [[] for _ in range(v + 1)]
    
    for _ in range(e):
        a, b, c = map(int, input().split())
        graph[a].append((b, c))
        graph[b].append((a, c))
        
    s, t = map(int, input().split())
    costs = dijkstra(s, graph, v)
    print(costs[t])

if __name__ == "__main__":
    main()

File: 백준/Gold/14284. 간선 이어가기 2/간선 이어가기 2.java
import java.io.IOException;
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.StringTokenizer;
import java.util.List;
import java.util.Arrays;
import java.util.ArrayList;
import java.util.PriorityQueue;


class Node implements Comparable<Node> {
    int vertex, cost;
    
    public Node(int vertex, int cost) {
        this.vertex = vertex;
        this.cost = cost;
    }
    
    @Override
    public int compareTo(Node o) {
        return Integer.compare(this.cost, o.cost);
    }
}

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        
        int v = Integer.parseInt(st.nextToken());
        int e = Integer.parseInt(st.nextToken());
        
        List<List<Node>> graph = new ArrayList<>();
        for (int i = 0; i < v + 1; i++) {
            graph.add(new ArrayList<>());
        }
        
        for (int i = 0; i < e; i++) {
            st = new StringTokenizer(br.readLine());
            int a = Integer.parseInt(st.nextToken());
            int b = Integer.parseInt(st.nextToken());
            int c = Integer.parseInt(st.nextToken());
            graph.get(a).add(new Node(b, c));
            graph.get(b).add(new Node(a, c));
        }
        
        st = new StringTokenizer(br.readLine());
        int s = Integer.parseInt(st.nextToken());
        int t = Integer.parseInt(st.nextToken());
        
        System.out.println(dijkstra(s, t, v, graph));
    }
    
    static int dijkstra(int start, int target, int n, List<List<Node>> graph) {
        int[] costs = new int[n + 1];
        Arrays.fill(costs, Integer.MAX_VALUE);
        PriorityQueue<Node> pQ = new PriorityQueue<>();
        
        costs[start] = 0; 
        pQ.add(new Node(start, 0));
        
        while (!pQ.isEmpty()) {
            Node cur = pQ.poll();
            
            int curNode = cur.vertex;
            int curCost = cur.cost;
            if (curCost > costs[curNode]) continue; 
            
            for (Node neighbor: graph.get(curNode)) {
                int newCost = curCost + neighbor.cost;
                
                if (newCost < costs[neighbor.vertex]) {
                    costs[neighbor.vertex] = newCost;
                    pQ.add(new Node(neighbor.vertex, newCost));
                }
            }
        }
        return costs[target];
    }
}